### PR TITLE
fix(TabBar):  Event forwarding fails when hosted in a ShadowRoot

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

### CODE REVIEW

File: packages/widgets/src/docklayout.ts [hunk 1]
```diff
@@ -1181,7 +1190,7 @@ export namespace DockLayout {
      *
      * @returns A new tab bar for a dock layout.
      */
-    createTabBar(): TabBar<Widget>;
+    createTabBar(document: Document | ShadowRoot): TabBar<Widget>;
```
Comment (blink1073) [inline packages/widgets/src/docklayout.ts:1193]: I think this argument should be optional

File: packages/widgets/src/tabbar.ts [hunk 2]
```diff
@@ -156,6 +157,13 @@ export class TabBar<T> extends Widget {
    */
   readonly renderer: TabBar.IRenderer<T>;
 
+  /**
+   * The document to use with the tab bar.
+   *
+   * The default is the global `document` instance.
+   */
+  document: Document | ShadowRoot;
```
Comment (blink1073) [inline packages/widgets/src/tabbar.ts:165]: I think this should be readonly, and only settable in the constructor.

**Review by blink1073 — CHANGES_REQUESTED** [review]
Thanks!  Just a couple of suggestions.

File: packages/widgets/src/docklayout.ts (hunk 1 above)
Comment (GordonSmith) [reply packages/widgets/src/docklayout.ts]: Change made as suggested  (FWIW it only actually gets called from one place, within the DockPanel),

File: packages/widgets/src/tabbar.ts (hunk 2 above)
Comment (GordonSmith) [reply packages/widgets/src/tabbar.ts]: Change made as suggested.

**Comment by GordonSmith** [comment]
@blink1073 FYI these changes are needed so that lumino widgets can be used within custom web components, which in turn can expose lumino widgets via "simple" html, see POC repo here (not currently published):  
* https://gordonsmith.github.io/hpcc-js-wc/components/splitter.html
* https://gordonsmith.github.io/hpcc-js-wc/components/dockpanel.html

This should greatly simplify the use of lumino widgets within frameworks like React, Vue and Angular.

**Comment by blink1073** [comment]
> This should greatly simplify the use of lumino widgets within frameworks like React, Vue and Angular.

Very nice!

**Comment by blink1073** [comment]
@GordonSmith, are you ready for me to make a release?

**Comment by GordonSmith** [comment]
@blink1073 no real rush as I am building off local branches for now.

**Comment by GordonSmith** [comment]
@blink1073 I might request a build if your not planning to do a release for a while?

Did you have any thoughts on the above examples?

**Comment by blink1073** [comment]
The examples are cool!  Happy to feature them as part of the https://github.com/jupyterlab/lumino#external-examples if you want to submit a PR.  I'll cut a release on Monday (I don't like to make releases on a Friday).

File: packages/widgets/src/docklayout.ts [hunk 3]
```diff
@@ -2184,7 +2193,7 @@ namespace Private {
     renderer: DockLayout.IRenderer
   ): TabLayoutNode {
     // Create the tab bar for the layout node.
-    let tabBar = renderer.createTabBar();
+    let tabBar = renderer.createTabBar(this._document);
```
Comment (hbcarlos) [inline packages/widgets/src/docklayout.ts:2196]: Does this exist in this context?
This function is inside the Private namespace.

Comment (hbcarlos) [reply packages/widgets/src/docklayout.ts:2196]: I'm updating lumino in JupyterLab, and there is an issue with `this._document`, when changing to a single-mode document.
https://github.com/jupyterlab/jupyterlab/pull/11823

Comment (GordonSmith) [reply packages/widgets/src/docklayout.ts:2196]: Do you have an example of the breaking code - I couldn't see anything related in the failing visual regression test?

Comment (GordonSmith) [reply packages/widgets/src/docklayout.ts:2196]: Fix incomming.

Comment (GordonSmith) [reply packages/widgets/src/docklayout.ts:2196]: FYI:  https://github.com/jupyterlab/lumino/pull/281

Comment (hbcarlos) [reply packages/widgets/src/docklayout.ts:2196]: Hey @GordonSmith, thanks for the quick reply.

> Do you have an example of the breaking code

Yes, when in JupyterLab you change the view to a single-mode document, there is an error on this line:
https://github.com/jupyterlab/jupyterlab/blob/3b21d568035f7383dfd88332e051277130aea7d8/packages/application/src/shell.ts#L557

Following the trace, you can see that the error comes from
https://github.com/jupyterlab/lumino/blob/948eec9353d06ff9808dc127c888a49a9930ca75/packages/widgets/src/docklayout.ts#L2196

where `this` is null.

**Pull request merged**